{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gzip\n",
    "import json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To parse data from json and info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def parse_data(info_path, json_zip_path):\n",
    "   # loads data\n",
    "   print(f\"Loading {json_zip_path}...\")\n",
    "   with gzip.open(json_zip_path, \"r\") as f:\n",
    "      data = [json.loads(line) for line in f]\n",
    "   \n",
    "   # loads data with label\n",
    "   print(f\"Loading {info_path}...\")\n",
    "   info = pd.read_csv(info_path)\n",
    "\n",
    "   #transfer information from json dict to list\n",
    "   print(\"Transferring data from json to dataframe...\")\n",
    "   res = []\n",
    "   for row in data:\n",
    "      for trans_id in row.keys():\n",
    "         for trans_pos in row[trans_id].keys():\n",
    "            for nucleo_seq in row[trans_id][trans_pos].keys():\n",
    "               temp = list(np.mean(np.array(row[trans_id][trans_pos][nucleo_seq]), axis=0))\n",
    "               # to get raw data without aggregation\n",
    "               # for features in row[trans_id][trans_pos][nucleo_seq]:\n",
    "               res.append([trans_id, int(trans_pos), nucleo_seq] + temp)\n",
    "\n",
    "   data = pd.DataFrame(res, columns = ['transcript_id', 'transcript_pos', 'nucleo_seq',\n",
    "                                       'dwell_time_-1', 'sd_-1', 'mean_-1',\n",
    "                                       'dwell_time_0', 'sd_0', 'mean_0',\n",
    "                                       'dwell_time_1', 'sd_1', 'mean_1'\n",
    "                                       ])\n",
    "   # Merge json data with labels\n",
    "   print(\"Merging dataframes to obtain labels\")\n",
    "   merged_data = pd.merge(data,info,left_on=[\"transcript_id\",\"transcript_pos\"],right_on=[\"transcript_id\",\"transcript_position\"], how=\"left\")\n",
    "\n",
    "   return merged_data\n",
    "\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /Users/shaun/Desktop/DSA4266/project2/dataset0.json.gz...\n",
      "Loading /Users/shaun/Desktop/DSA4266/project2/data.info...\n",
      "Transferring data from json to dataframe...\n",
      "Merging dataframes to obtain labels\n"
     ]
    }
   ],
   "source": [
    "data = parse_data(\"/Users/shaun/Desktop/DSA4266/project2/data.info\",'/Users/shaun/Desktop/DSA4266/project2/dataset0.json.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_pos</th>\n",
       "      <th>nucleo_seq</th>\n",
       "      <th>dwell_time_-1</th>\n",
       "      <th>sd_-1</th>\n",
       "      <th>mean_-1</th>\n",
       "      <th>dwell_time_0</th>\n",
       "      <th>sd_0</th>\n",
       "      <th>mean_0</th>\n",
       "      <th>dwell_time_1</th>\n",
       "      <th>sd_1</th>\n",
       "      <th>mean_1</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>0.007345</td>\n",
       "      <td>4.386989</td>\n",
       "      <td>80.570270</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>244</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>0.007710</td>\n",
       "      <td>3.016599</td>\n",
       "      <td>94.290698</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>261</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>0.007555</td>\n",
       "      <td>2.087146</td>\n",
       "      <td>89.364324</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>316</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>332</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>0.006102</td>\n",
       "      <td>2.236520</td>\n",
       "      <td>89.154000</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>332</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>368</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>0.010019</td>\n",
       "      <td>4.260253</td>\n",
       "      <td>85.178788</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>368</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1348</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>0.009594</td>\n",
       "      <td>3.294164</td>\n",
       "      <td>118.232877</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>4.929726</td>\n",
       "      <td>116.342466</td>\n",
       "      <td>0.006555</td>\n",
       "      <td>4.005616</td>\n",
       "      <td>82.004110</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>1348</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1429</td>\n",
       "      <td>CTGACAC</td>\n",
       "      <td>0.008393</td>\n",
       "      <td>4.511014</td>\n",
       "      <td>110.969565</td>\n",
       "      <td>0.010305</td>\n",
       "      <td>9.105797</td>\n",
       "      <td>114.927536</td>\n",
       "      <td>0.005568</td>\n",
       "      <td>3.644638</td>\n",
       "      <td>80.497101</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>1429</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1531</td>\n",
       "      <td>TGGACAC</td>\n",
       "      <td>0.008161</td>\n",
       "      <td>3.918438</td>\n",
       "      <td>113.968750</td>\n",
       "      <td>0.006877</td>\n",
       "      <td>4.759687</td>\n",
       "      <td>113.562500</td>\n",
       "      <td>0.006410</td>\n",
       "      <td>2.181562</td>\n",
       "      <td>84.190625</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>1531</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1537</td>\n",
       "      <td>CTGACCA</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>3.191228</td>\n",
       "      <td>109.354386</td>\n",
       "      <td>0.007419</td>\n",
       "      <td>6.552982</td>\n",
       "      <td>123.263158</td>\n",
       "      <td>0.006472</td>\n",
       "      <td>2.540877</td>\n",
       "      <td>82.289474</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>1537</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.008788</td>\n",
       "      <td>4.090577</td>\n",
       "      <td>105.807692</td>\n",
       "      <td>0.006907</td>\n",
       "      <td>8.702885</td>\n",
       "      <td>113.134615</td>\n",
       "      <td>0.008337</td>\n",
       "      <td>2.576731</td>\n",
       "      <td>78.536538</td>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>1693</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          transcript_id  transcript_pos nucleo_seq  dwell_time_-1     sd_-1  \\\n",
       "0       ENST00000000233             244    AAGACCA       0.008264  4.223784   \n",
       "1       ENST00000000233             261    CAAACTG       0.006609  3.216424   \n",
       "2       ENST00000000233             316    GAAACAG       0.007570  2.940541   \n",
       "3       ENST00000000233             332    AGAACAT       0.010620  6.476350   \n",
       "4       ENST00000000233             368    AGGACAA       0.010701  6.415051   \n",
       "...                 ...             ...        ...            ...       ...   \n",
       "121833  ENST00000641834            1348    GGGACAT       0.009594  3.294164   \n",
       "121834  ENST00000641834            1429    CTGACAC       0.008393  4.511014   \n",
       "121835  ENST00000641834            1531    TGGACAC       0.008161  3.918438   \n",
       "121836  ENST00000641834            1537    CTGACCA       0.008044  3.191228   \n",
       "121837  ENST00000641834            1693    TTGACAT       0.008788  4.090577   \n",
       "\n",
       "           mean_-1  dwell_time_0      sd_0      mean_0  dwell_time_1  \\\n",
       "0       123.702703      0.009373  7.382162  125.913514      0.007345   \n",
       "1       109.681395      0.006813  3.226535  107.889535      0.007710   \n",
       "2       105.475676      0.007416  3.642703   98.947027      0.007555   \n",
       "3       129.355000      0.008632  2.899200   97.836500      0.006102   \n",
       "4       117.924242      0.011479  5.870303  121.954545      0.010019   \n",
       "...            ...           ...       ...         ...           ...   \n",
       "121833  118.232877      0.007300  4.929726  116.342466      0.006555   \n",
       "121834  110.969565      0.010305  9.105797  114.927536      0.005568   \n",
       "121835  113.968750      0.006877  4.759687  113.562500      0.006410   \n",
       "121836  109.354386      0.007419  6.552982  123.263158      0.006472   \n",
       "121837  105.807692      0.006907  8.702885  113.134615      0.008337   \n",
       "\n",
       "            sd_1     mean_1          gene_id  transcript_position  label  \n",
       "0       4.386989  80.570270  ENSG00000004059                  244      0  \n",
       "1       3.016599  94.290698  ENSG00000004059                  261      0  \n",
       "2       2.087146  89.364324  ENSG00000004059                  316      0  \n",
       "3       2.236520  89.154000  ENSG00000004059                  332      0  \n",
       "4       4.260253  85.178788  ENSG00000004059                  368      0  \n",
       "...          ...        ...              ...                  ...    ...  \n",
       "121833  4.005616  82.004110  ENSG00000167747                 1348      1  \n",
       "121834  3.644638  80.497101  ENSG00000167747                 1429      0  \n",
       "121835  2.181562  84.190625  ENSG00000167747                 1531      1  \n",
       "121836  2.540877  82.289474  ENSG00000167747                 1537      0  \n",
       "121837  2.576731  78.536538  ENSG00000167747                 1693      0  \n",
       "\n",
       "[121838 rows x 15 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Feature engineering "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One Hot Encoding of the sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def onehote(sequence):\n",
    "    mapping = {\"A\": 0, \"C\": 1, \"G\": 2, \"T\": 3}\n",
    "    seq2 = [mapping[i] for i in sequence]\n",
    "    return np.eye(4)[seq2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get Kmer sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kmers(sequence, ksize=5):\n",
    "    kmers = []\n",
    "    n_kmers = len(sequence) - ksize + 1\n",
    "\n",
    "    for i in range(n_kmers):\n",
    "        kmer = sequence[i:i + ksize]\n",
    "        kmers.append(kmer)\n",
    "\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get position weight matrix score for each nucleo_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_pwm_from_sequences(sequences):\n",
    "    \"\"\"\n",
    "    Create a Position Weight Matrix (PWM) from a list of sequences.\n",
    "    \n",
    "    Args:\n",
    "        sequences (list of str): List of DNA sequences of equal length.\n",
    "        \n",
    "    Returns:\n",
    "        np.ndarray: The PWM matrix.\n",
    "    \"\"\"\n",
    "    # Check if all sequences have the same length\n",
    "    seq_length = len(sequences[0])\n",
    "    if not all(len(seq) == seq_length for seq in sequences):\n",
    "        raise ValueError(\"All sequences must have the same length\")\n",
    "\n",
    "    # Initialize the PWM matrix with zeros\n",
    "    pwm = np.zeros((4, seq_length))  # 4 rows for A, C, G, T\n",
    "\n",
    "    # Count the occurrences of each nucleotide at each position\n",
    "    for seq in sequences:\n",
    "        for i, nucleotide in enumerate(seq):\n",
    "            if nucleotide == 'A':\n",
    "                pwm[0, i] += 1\n",
    "            elif nucleotide == 'C':\n",
    "                pwm[1, i] += 1\n",
    "            elif nucleotide == 'G':\n",
    "                pwm[2, i] += 1\n",
    "            elif nucleotide == 'T':\n",
    "                pwm[3, i] += 1\n",
    "\n",
    "    # Convert counts to probabilities\n",
    "    pwm = pwm / len(sequences)\n",
    "\n",
    "    return pd.DataFrame(data=pwm, columns=[1,2,3,4,5,6,7], index=[\"A\",\"C\",\"T\",\"G\"])\n",
    "\n",
    "def log_odds(x):\n",
    "    \"\"\"\n",
    "    Function to calculate log odds\n",
    "    \"\"\"\n",
    "    if x == 0:\n",
    "        return 0\n",
    "    else:\n",
    "        return np.log2(x/0.25)\n",
    "\n",
    "\n",
    "def get_log_odds(sequences):\n",
    "    \"\"\"To get log odd dictionary which is used to calculate PWM score for each nucleo_seq\"\"\"\n",
    "    ppm = create_pwm_from_sequences(sequences)\n",
    "    log_odds_pos = ppm.applymap(log_odds)\n",
    "    log_odds_dict = log_odds_pos.to_dict()\n",
    "\n",
    "    return log_odds_dict\n",
    "\n",
    "def get_PWM_score(seq):\n",
    "    \"\"\"Calculating PWM score for nucleo_seq based on the log odds\"\"\"\n",
    "    res = 0\n",
    "    for i in range(len(seq)):\n",
    "        base = seq[i]\n",
    "        dic = log_odds_dict[(i+1)]\n",
    "        res = res + dic[base]\n",
    "    return res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmer_df = pd.DataFrame(data.apply(lambda x: kmers(x.nucleo_seq),axis=1).to_list(), columns=[\"nucleo_seq_-1\",\"nucleo_seq_0\",\"nucleo_seq_1\"])\n",
    "data = pd.concat([data,kmer_df], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "data[\"seq\"] = data.apply(lambda x: onehote(x.nucleo_seq), axis=1)\n",
    "data[\"seq_-1\"] = data.apply(lambda x: onehote(x[\"nucleo_seq_-1\"]), axis=1)\n",
    "data[\"seq_0\"] = data.apply(lambda x: onehote(x.nucleo_seq_0), axis=1)\n",
    "data[\"seq_1\"] = data.apply(lambda x: onehote(x.nucleo_seq_1), axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_odds_dict = get_log_odds(data.nucleo_seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "data[\"pwm_score\"] = data.apply(lambda x: get_PWM_score(x.nucleo_seq),axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_pos</th>\n",
       "      <th>nucleo_seq</th>\n",
       "      <th>dwell_time_-1</th>\n",
       "      <th>sd_-1</th>\n",
       "      <th>mean_-1</th>\n",
       "      <th>dwell_time_0</th>\n",
       "      <th>sd_0</th>\n",
       "      <th>mean_0</th>\n",
       "      <th>dwell_time_1</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>label</th>\n",
       "      <th>nucleo_seq_-1</th>\n",
       "      <th>nucleo_seq_0</th>\n",
       "      <th>nucleo_seq_1</th>\n",
       "      <th>seq</th>\n",
       "      <th>seq_-1</th>\n",
       "      <th>seq_0</th>\n",
       "      <th>seq_1</th>\n",
       "      <th>pwm_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGACCA</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>4.223784</td>\n",
       "      <td>123.702703</td>\n",
       "      <td>0.009373</td>\n",
       "      <td>7.382162</td>\n",
       "      <td>125.913514</td>\n",
       "      <td>0.007345</td>\n",
       "      <td>...</td>\n",
       "      <td>244</td>\n",
       "      <td>0</td>\n",
       "      <td>AAGAC</td>\n",
       "      <td>AGACC</td>\n",
       "      <td>GACCA</td>\n",
       "      <td>[[1.0, 0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>[[1.0, 0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>[[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>5.265638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>CAAACTG</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>3.216424</td>\n",
       "      <td>109.681395</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>3.226535</td>\n",
       "      <td>107.889535</td>\n",
       "      <td>0.007710</td>\n",
       "      <td>...</td>\n",
       "      <td>261</td>\n",
       "      <td>0</td>\n",
       "      <td>CAAAC</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>AACTG</td>\n",
       "      <td>[[0.0, 1.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>[[0.0, 1.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>[[1.0, 0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>[[1.0, 0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>5.282447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>GAAACAG</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>2.940541</td>\n",
       "      <td>105.475676</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>3.642703</td>\n",
       "      <td>98.947027</td>\n",
       "      <td>0.007555</td>\n",
       "      <td>...</td>\n",
       "      <td>316</td>\n",
       "      <td>0</td>\n",
       "      <td>GAAAC</td>\n",
       "      <td>AAACA</td>\n",
       "      <td>AACAG</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>[[1.0, 0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>[[1.0, 0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>6.220012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>332</td>\n",
       "      <td>AGAACAT</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>6.476350</td>\n",
       "      <td>129.355000</td>\n",
       "      <td>0.008632</td>\n",
       "      <td>2.899200</td>\n",
       "      <td>97.836500</td>\n",
       "      <td>0.006102</td>\n",
       "      <td>...</td>\n",
       "      <td>332</td>\n",
       "      <td>0</td>\n",
       "      <td>AGAAC</td>\n",
       "      <td>GAACA</td>\n",
       "      <td>AACAT</td>\n",
       "      <td>[[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...</td>\n",
       "      <td>[[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>[[1.0, 0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>5.970300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>368</td>\n",
       "      <td>AGGACAA</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>6.415051</td>\n",
       "      <td>117.924242</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>5.870303</td>\n",
       "      <td>121.954545</td>\n",
       "      <td>0.010019</td>\n",
       "      <td>...</td>\n",
       "      <td>368</td>\n",
       "      <td>0</td>\n",
       "      <td>AGGAC</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>GACAA</td>\n",
       "      <td>[[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...</td>\n",
       "      <td>[[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>5.146423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1348</td>\n",
       "      <td>GGGACAT</td>\n",
       "      <td>0.009594</td>\n",
       "      <td>3.294164</td>\n",
       "      <td>118.232877</td>\n",
       "      <td>0.007300</td>\n",
       "      <td>4.929726</td>\n",
       "      <td>116.342466</td>\n",
       "      <td>0.006555</td>\n",
       "      <td>...</td>\n",
       "      <td>1348</td>\n",
       "      <td>1</td>\n",
       "      <td>GGGAC</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>GACAT</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>4.610474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1429</td>\n",
       "      <td>CTGACAC</td>\n",
       "      <td>0.008393</td>\n",
       "      <td>4.511014</td>\n",
       "      <td>110.969565</td>\n",
       "      <td>0.010305</td>\n",
       "      <td>9.105797</td>\n",
       "      <td>114.927536</td>\n",
       "      <td>0.005568</td>\n",
       "      <td>...</td>\n",
       "      <td>1429</td>\n",
       "      <td>0</td>\n",
       "      <td>CTGAC</td>\n",
       "      <td>TGACA</td>\n",
       "      <td>GACAC</td>\n",
       "      <td>[[0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 1.0], [...</td>\n",
       "      <td>[[0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 1.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 1.0, 0.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>4.372075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1531</td>\n",
       "      <td>TGGACAC</td>\n",
       "      <td>0.008161</td>\n",
       "      <td>3.918438</td>\n",
       "      <td>113.968750</td>\n",
       "      <td>0.006877</td>\n",
       "      <td>4.759687</td>\n",
       "      <td>113.562500</td>\n",
       "      <td>0.006410</td>\n",
       "      <td>...</td>\n",
       "      <td>1531</td>\n",
       "      <td>1</td>\n",
       "      <td>TGGAC</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>GACAC</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 1.0, 0.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 1.0, 0.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>4.548100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1537</td>\n",
       "      <td>CTGACCA</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>3.191228</td>\n",
       "      <td>109.354386</td>\n",
       "      <td>0.007419</td>\n",
       "      <td>6.552982</td>\n",
       "      <td>123.263158</td>\n",
       "      <td>0.006472</td>\n",
       "      <td>...</td>\n",
       "      <td>1537</td>\n",
       "      <td>0</td>\n",
       "      <td>CTGAC</td>\n",
       "      <td>TGACC</td>\n",
       "      <td>GACCA</td>\n",
       "      <td>[[0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 1.0], [...</td>\n",
       "      <td>[[0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 1.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 1.0, 0.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>4.355071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>TTGACAT</td>\n",
       "      <td>0.008788</td>\n",
       "      <td>4.090577</td>\n",
       "      <td>105.807692</td>\n",
       "      <td>0.006907</td>\n",
       "      <td>8.702885</td>\n",
       "      <td>113.134615</td>\n",
       "      <td>0.008337</td>\n",
       "      <td>...</td>\n",
       "      <td>1693</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGAC</td>\n",
       "      <td>TGACA</td>\n",
       "      <td>GACAT</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 0.0, 1.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 0.0, 1.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 1.0, 0.0], [...</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...</td>\n",
       "      <td>5.054085</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          transcript_id  transcript_pos nucleo_seq  dwell_time_-1     sd_-1  \\\n",
       "0       ENST00000000233             244    AAGACCA       0.008264  4.223784   \n",
       "1       ENST00000000233             261    CAAACTG       0.006609  3.216424   \n",
       "2       ENST00000000233             316    GAAACAG       0.007570  2.940541   \n",
       "3       ENST00000000233             332    AGAACAT       0.010620  6.476350   \n",
       "4       ENST00000000233             368    AGGACAA       0.010701  6.415051   \n",
       "...                 ...             ...        ...            ...       ...   \n",
       "121833  ENST00000641834            1348    GGGACAT       0.009594  3.294164   \n",
       "121834  ENST00000641834            1429    CTGACAC       0.008393  4.511014   \n",
       "121835  ENST00000641834            1531    TGGACAC       0.008161  3.918438   \n",
       "121836  ENST00000641834            1537    CTGACCA       0.008044  3.191228   \n",
       "121837  ENST00000641834            1693    TTGACAT       0.008788  4.090577   \n",
       "\n",
       "           mean_-1  dwell_time_0      sd_0      mean_0  dwell_time_1  ...  \\\n",
       "0       123.702703      0.009373  7.382162  125.913514      0.007345  ...   \n",
       "1       109.681395      0.006813  3.226535  107.889535      0.007710  ...   \n",
       "2       105.475676      0.007416  3.642703   98.947027      0.007555  ...   \n",
       "3       129.355000      0.008632  2.899200   97.836500      0.006102  ...   \n",
       "4       117.924242      0.011479  5.870303  121.954545      0.010019  ...   \n",
       "...            ...           ...       ...         ...           ...  ...   \n",
       "121833  118.232877      0.007300  4.929726  116.342466      0.006555  ...   \n",
       "121834  110.969565      0.010305  9.105797  114.927536      0.005568  ...   \n",
       "121835  113.968750      0.006877  4.759687  113.562500      0.006410  ...   \n",
       "121836  109.354386      0.007419  6.552982  123.263158      0.006472  ...   \n",
       "121837  105.807692      0.006907  8.702885  113.134615      0.008337  ...   \n",
       "\n",
       "        transcript_position  label nucleo_seq_-1  nucleo_seq_0  nucleo_seq_1  \\\n",
       "0                       244      0         AAGAC         AGACC         GACCA   \n",
       "1                       261      0         CAAAC         AAACT         AACTG   \n",
       "2                       316      0         GAAAC         AAACA         AACAG   \n",
       "3                       332      0         AGAAC         GAACA         AACAT   \n",
       "4                       368      0         AGGAC         GGACA         GACAA   \n",
       "...                     ...    ...           ...           ...           ...   \n",
       "121833                 1348      1         GGGAC         GGACA         GACAT   \n",
       "121834                 1429      0         CTGAC         TGACA         GACAC   \n",
       "121835                 1531      1         TGGAC         GGACA         GACAC   \n",
       "121836                 1537      0         CTGAC         TGACC         GACCA   \n",
       "121837                 1693      0         TTGAC         TGACA         GACAT   \n",
       "\n",
       "                                                      seq  \\\n",
       "0       [[1.0, 0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...   \n",
       "1       [[0.0, 1.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...   \n",
       "2       [[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...   \n",
       "3       [[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...   \n",
       "4       [[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...   \n",
       "...                                                   ...   \n",
       "121833  [[0.0, 0.0, 1.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...   \n",
       "121834  [[0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 1.0], [...   \n",
       "121835  [[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 1.0, 0.0], [...   \n",
       "121836  [[0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 1.0], [...   \n",
       "121837  [[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 0.0, 1.0], [...   \n",
       "\n",
       "                                                   seq_-1  \\\n",
       "0       [[1.0, 0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...   \n",
       "1       [[0.0, 1.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...   \n",
       "2       [[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...   \n",
       "3       [[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...   \n",
       "4       [[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...   \n",
       "...                                                   ...   \n",
       "121833  [[0.0, 0.0, 1.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...   \n",
       "121834  [[0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 1.0], [...   \n",
       "121835  [[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 1.0, 0.0], [...   \n",
       "121836  [[0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 1.0], [...   \n",
       "121837  [[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 0.0, 1.0], [...   \n",
       "\n",
       "                                                    seq_0  \\\n",
       "0       [[1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...   \n",
       "1       [[1.0, 0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...   \n",
       "2       [[1.0, 0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...   \n",
       "3       [[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...   \n",
       "4       [[0.0, 0.0, 1.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...   \n",
       "...                                                   ...   \n",
       "121833  [[0.0, 0.0, 1.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...   \n",
       "121834  [[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 1.0, 0.0], [...   \n",
       "121835  [[0.0, 0.0, 1.0, 0.0], [0.0, 0.0, 1.0, 0.0], [...   \n",
       "121836  [[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 1.0, 0.0], [...   \n",
       "121837  [[0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 1.0, 0.0], [...   \n",
       "\n",
       "                                                    seq_1 pwm_score  \n",
       "0       [[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...  5.265638  \n",
       "1       [[1.0, 0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...  5.282447  \n",
       "2       [[1.0, 0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...  6.220012  \n",
       "3       [[1.0, 0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...  5.970300  \n",
       "4       [[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...  5.146423  \n",
       "...                                                   ...       ...  \n",
       "121833  [[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...  4.610474  \n",
       "121834  [[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...  4.372075  \n",
       "121835  [[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...  4.548100  \n",
       "121836  [[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...  4.355071  \n",
       "121837  [[0.0, 0.0, 1.0, 0.0], [1.0, 0.0, 0.0, 0.0], [...  5.054085  \n",
       "\n",
       "[121838 rows x 23 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Train test split by gene ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import GroupShuffleSplit \n",
    "\n",
    "splitter = GroupShuffleSplit(test_size=.10, n_splits=2, random_state = 42)\n",
    "split = splitter.split(data, groups=data['gene_id'])\n",
    "train_idx, test_idx = next(split)\n",
    "\n",
    "train = data.iloc[train_idx]\n",
    "test = data.iloc[test_idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_pos</th>\n",
       "      <th>nucleo_seq</th>\n",
       "      <th>dwell_time_-1</th>\n",
       "      <th>sd_-1</th>\n",
       "      <th>mean_-1</th>\n",
       "      <th>dwell_time_0</th>\n",
       "      <th>sd_0</th>\n",
       "      <th>mean_0</th>\n",
       "      <th>dwell_time_1</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_position</th>\n",
       "      <th>label</th>\n",
       "      <th>nucleo_seq_-1</th>\n",
       "      <th>nucleo_seq_0</th>\n",
       "      <th>nucleo_seq_1</th>\n",
       "      <th>seq</th>\n",
       "      <th>seq_-1</th>\n",
       "      <th>seq_0</th>\n",
       "      <th>seq_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [transcript_id, transcript_pos, nucleo_seq, dwell_time_-1, sd_-1, mean_-1, dwell_time_0, sd_0, mean_0, dwell_time_1, sd_1, mean_1, gene_id, transcript_position, label, nucleo_seq_-1, nucleo_seq_0, nucleo_seq_1, seq, seq_-1, seq_0, seq_1]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 22 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train[train[\"gene_id\"] == \"ENSG00000006451\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "e5ed79da61d64e3c326f85804a9a2c1fdc2c426a0b478f8b6c1cd1827151984a"
  },
  "kernelspec": {
   "display_name": "Python 3.8.3 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
